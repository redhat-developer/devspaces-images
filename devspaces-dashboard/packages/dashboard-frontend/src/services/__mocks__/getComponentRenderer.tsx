/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { render } from '@testing-library/react';
import React from 'react';
import renderer from 'react-test-renderer';

export * from '@testing-library/react';

export default function <T extends Array<unknown>>(
  getComponent: (..._args: T) => React.ReactElement,
): {
  createSnapshot: (...args: T) => renderer.ReactTestRenderer;
  renderComponent: (...args: T) => {
    reRenderComponent: (..._args: T) => void;
  };
} {
  return {
    createSnapshot: (...args: T) => {
      return renderer.create(getComponent(...args));
    },
    renderComponent: (...args: T) => {
      const res = render(getComponent(...args));
      return {
        reRenderComponent: (..._args: T) => {
          res.rerender(getComponent(..._args));
        },
      };
    },
  };
}
