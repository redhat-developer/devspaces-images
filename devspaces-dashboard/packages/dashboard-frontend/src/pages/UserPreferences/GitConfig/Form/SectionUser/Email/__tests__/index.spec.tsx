/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { ValidatedOptions } from '@patternfly/react-core';
import userEvent from '@testing-library/user-event';
import * as React from 'react';

import getComponentRenderer, { screen } from '@/services/__mocks__/getComponentRenderer';

import { GitConfigUserEmail } from '..';

jest.mock('@/components/InputGroupExtended');

const { createSnapshot, renderComponent } = getComponentRenderer(getComponent);

const mockOnChange = jest.fn();

describe('GitConfigUserEmail', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  test('snapshot, not loading', () => {
    const snapshot = createSnapshot('user@che', false);
    expect(snapshot.toJSON()).toMatchSnapshot();
  });
  test('snapshot', () => {
    const snapshot = createSnapshot('user@che', true);
    expect(snapshot.toJSON()).toMatchSnapshot();
  });

  it('should fail validation if value is empty', async () => {
    renderComponent('user@che.org', false);

    const textInput = screen.getByRole('textbox');
    await userEvent.clear(textInput);

    expect(screen.getByTestId('validated')).toHaveTextContent(ValidatedOptions.error);
  });

  it('should fail validation if value is too long', async () => {
    renderComponent('user@che.org', false);

    const textInput = screen.getByRole('textbox');
    await userEvent.click(textInput);
    await userEvent.paste('a'.repeat(129));

    expect(screen.getByTestId('validated')).toHaveTextContent(ValidatedOptions.error);
  });

  it('should fail validation if is not a valid email', async () => {
    renderComponent('user@che.org', false);

    const textInput = screen.getByRole('textbox');
    await userEvent.click(textInput);
    await userEvent.paste('@test');

    expect(screen.getByTestId('validated')).toHaveTextContent(ValidatedOptions.error);
  });

  it('should re-validate on component update', () => {
    const { reRenderComponent } = renderComponent('', false);

    expect(screen.getByTestId('validated')).toHaveTextContent(ValidatedOptions.error);

    reRenderComponent('user@che.com', false);

    expect(screen.getByTestId('validated')).toHaveTextContent(ValidatedOptions.default);
  });

  it('should handle value changing', async () => {
    renderComponent('user@che.org', false);

    const textInput = screen.getByRole('textbox');
    await userEvent.click(textInput);
    await userEvent.paste('a');

    expect(mockOnChange).toHaveBeenCalledWith('user@che.orga', true);
  });
});

function getComponent(value: string, isLoading: boolean): React.ReactElement {
  return <GitConfigUserEmail isLoading={isLoading} value={value} onChange={mockOnChange} />;
}
