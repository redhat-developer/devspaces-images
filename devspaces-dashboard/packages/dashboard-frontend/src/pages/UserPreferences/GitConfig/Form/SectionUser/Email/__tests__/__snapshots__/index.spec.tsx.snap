// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GitConfigUserEmail snapshot 1`] = `
<div
  className="pf-c-form__group"
>
  <div
    className="pf-c-form__group-label"
  >
    <label
      className="pf-c-form__label"
      htmlFor="gitconfig-user-email"
    >
      <span
        className="pf-c-form__label-text"
      >
        email
      </span>
      <span
        aria-hidden="true"
        className="pf-c-form__label-required"
      >
         
        *
      </span>
    </label>
     
  </div>
  <div
    className="pf-c-form__group-control"
  >
    <div>
      <input
        aria-invalid={true}
        aria-label={null}
        className="pf-c-form-control"
        data-ouia-component-id="OUIA-Generated-TextInputBase-2"
        data-ouia-component-type="PF4/TextInput"
        data-ouia-safe={true}
        disabled={true}
        id="gitconfig-user-email"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        required={false}
        type="text"
        value="user@che"
      />
      <span
        data-testid="validated"
      >
        error
      </span>
      <button
        aria-disabled={false}
        aria-label={null}
        className="pf-c-button pf-m-primary"
        data-ouia-component-id="OUIA-Generated-Button-primary-2"
        data-ouia-component-type="PF4/Button"
        data-ouia-safe={true}
        data-testid="button-remove"
        disabled={false}
        role={null}
        type="button"
      />
    </div>
    <div
      aria-live="polite"
      className="pf-c-form__helper-text pf-m-error"
      id="gitconfig-user-email-helper"
    >
      The value is not a valid email address.
    </div>
  </div>
</div>
`;

exports[`GitConfigUserEmail snapshot, not loading 1`] = `
<div
  className="pf-c-form__group"
>
  <div
    className="pf-c-form__group-label"
  >
    <label
      className="pf-c-form__label"
      htmlFor="gitconfig-user-email"
    >
      <span
        className="pf-c-form__label-text"
      >
        email
      </span>
      <span
        aria-hidden="true"
        className="pf-c-form__label-required"
      >
         
        *
      </span>
    </label>
     
  </div>
  <div
    className="pf-c-form__group-control"
  >
    <div>
      <input
        aria-invalid={true}
        aria-label={null}
        className="pf-c-form-control"
        data-ouia-component-id="OUIA-Generated-TextInputBase-1"
        data-ouia-component-type="PF4/TextInput"
        data-ouia-safe={true}
        disabled={false}
        id="gitconfig-user-email"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        required={false}
        type="text"
        value="user@che"
      />
      <span
        data-testid="validated"
      >
        error
      </span>
      <button
        aria-disabled={false}
        aria-label={null}
        className="pf-c-button pf-m-primary"
        data-ouia-component-id="OUIA-Generated-Button-primary-1"
        data-ouia-component-type="PF4/Button"
        data-ouia-safe={true}
        data-testid="button-remove"
        disabled={false}
        role={null}
        type="button"
      />
    </div>
    <div
      aria-live="polite"
      className="pf-c-form__helper-text pf-m-error"
      id="gitconfig-user-email-helper"
    >
      The value is not a valid email address.
    </div>
  </div>
</div>
`;
